### PR TITLE
feat(core): add dependency injection for Database and LocalClient

### LINKED ISSUE #1649: Replace stack-based context in Command builder with pre/post hook calls
Command builder currently uses the same `stack.enter_context` code that was use in `pass_local_client`. We should replace this with proper pre-/post-hook implementations in the builders. We might need to improve on error handling of the current builder a bit for this. 

We should also try to improve cleanup where it makes sense, to not leave the repo in a dirty state on failure and not leak a lock.

### LINKED ISSUE #1654: Add typing information to Command builder
In an effort to improve renku python in general, we should start adding python typing information where possible. the command builder is a good starting place, with the goal of one day adding it everywhere.

### LINKED ISSUE #2109: Add CommandBuilder for Storage backend
We should have a new `CommandBuilder` command like `with_storage()` that, in the pre-hook opens the `Storage` for the local repo and makes it accessible where needed. This could be done similar to `.with_communicator(communicator)` that we already have, so it's injected in the relevant places and doesn't have to be passed around (pay attention to thread safety).

In the posthook of the CommandBuilder, it should call `Storage.commit()` so changes get stored (if there are any), maybe with a flag that prevents this in the case of read-only commands. This has to happen BEFORE we make a commit to git for a command Note: I think currently the commit for a command still happens using the old contextmanager -> with commit(): style code, not in a proper post hook. We should refactor that to use the prehook/posthook logic of CommandBuilder in this story as well.

### PR DESCRIPTION
**Pull request by Panaetius**
closes #2109
closes #1649
closes #1654

### CODE REVIEW

File: tests/utils.py [hunk 1]
```diff
@@ -105,3 +106,23 @@ def modified_environ(*remove, **update):
     finally:
         env.update(update_after)
         [env.pop(k) for k in remove_after]
+
+
+@contextmanager
+def inject_client(client):
```
Comment (mohammad-sdsc) [inline tests/utils.py:112]: This is not used anywhere else and can be deleted.

**Review by mohammad-sdsc — COMMENTED** [review]
Looks great! Thanks! Just made a few minor comments.

File: renku/core/management/command_builder/command.py [hunk 2]
```diff
@@ -49,37 +50,116 @@ def wrapper(*args, **kwargs):
...
+def _patched_configure(
+    config: typing.Optional[inject.BinderCallable] = None, bind_in_runtime: bool = True
+) -> inject.Injector:
+    """Create an injector with a callable config or raise an exception when already configured."""
+
+    if getattr(_LOCAL, "injector", None):
+        raise inject.InjectorException("Injector is already configured")
+
+    _LOCAL.injector = inject.Injector(config, bind_in_runtime=bind_in_runtime)
+    inject.logger.debug("Created and configured an injector, config=%s", config)
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/management/command_builder/command.py:74]: oh it's just copied over from the original code, but you're right, we don't really need it

**Review by mohammad-sdsc — APPROVED** [review]
Thank you!

**Pull request merged**